crates/bevy_ecs/src/traversal.rs

```diff
@@ -0,0 +1,39 @@
+//! A trait for components that let you traverse the ECS.
+
+use crate::{
+    component::{Component, StorageType},
+    entity::Entity,
+};
+
+/// A component that can point to another entity, and which can be used to define a path through the ECS.
+///
+/// Traversals are used to [specify the direction] of [event propagation] in [observers]. By default,
+/// events use the [`TraverseNone`] placeholder component, which cannot actually be created or added to
+/// an entity and so never causes traversal.
+///
+/// The implementer is responsible for ensuring that `Traversal::next` cannot produce infinite loops.
+///
+/// [specify the direction]: crate::event::Event::Traverse
+/// [event propagation]: crate::observer::Trigger::propagate
+/// [observers]: crate::observer::Observer
+pub trait Traversal: Component {
```

Comment:
So it will be highly likely that traversals will be along "relations at home" components. With hooks people can now ensure that these graphs never become invalid. However to do so underlying `Component` types have to be private otherwise you can break invariants by doing this:
```rs
let foo = world.entity(a).take::<Foo>().unwrap();
*world.entity(b).get_mut::<Foo>().unwrap() = foo;
```
I already do this in aery and hierarchy will have to do this when it starts using hooks so rather than being `Component` this should be `QueryData` and `next` should take the item instead of self. You can just remove `TraversalNone` and make the default query `()`.

Reply:
Ugh actually this change might require #13375 first.

Reply:
Yep, I tried to design it this way originally and couldn't get the ref safety right, then designed a "minimal" version to use a single component. This would be best addressed as early follow-up, but I totally want this.